<a href="https://colab.research.google.com/github/Tora0407/Challenge-TelecomX-Fase1/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [195]:
import pandas as pd
import numpy as np

## 1.1 Extracción de datos

In [196]:
df = pd.read_json('https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json')
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

#🔧 Transformación

#### Diccionario de datos

- `customerID`: número de identificación único de cada cliente
- `Churn`: si el cliente dejó o no la empresa
- `gender`: género (masculino y femenino)
- `SeniorCitizen`: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- `Partner`: si el cliente tiene o no una pareja
- `Dependents`: si el cliente tiene o no dependientes
- `tenure`: meses de contrato del cliente
- `PhoneService`: suscripción al servicio telefónico
- `MultipleLines`: suscripción a más de una línea telefónica
- `InternetService`: suscripción a un proveedor de internet
- `OnlineSecurity`: suscripción adicional de seguridad en línea
- `OnlineBackup`: suscripción adicional de respaldo en línea
- `DeviceProtection`: suscripción adicional de protección del dispositivo
- `TechSupport`: suscripción adicional de soporte técnico, menor tiempo de espera
- `StreamingTV`: suscripción de televisión por cable
- `StreamingMovies`: suscripción de streaming de películas
- `Contract`: tipo de contrato
- `PaperlessBilling`: si el cliente prefiere recibir la factura en línea
- `PaymentMethod`: forma de pago
- `Charges.Monthly`: total de todos los servicios del cliente por mes
- `Charges.Total`: total gastado por el cliente

## 2.1 Conoce el conjunto de datos

In [197]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


⚠️ Se observa que los datos se encuentran en diccionarios (`customer`, `phone`, `internet`, `account`), por lo tanto hay que normalizarlos

In [198]:
df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

#verificamos si se realizó la normalización
df_customer

gender  SeniorCitizen Partner Dependents  tenure
0     Female              0     Yes        Yes       9
1       Male              0      No         No       9
2       Male              0      No         No       4
3       Male              1     Yes         No      13
4     Female              1     Yes         No       3
...      ...            ...     ...        ...     ...
7262  Female              0      No         No      13
7263    Male              0     Yes         No      22
7264    Male              0      No         No       2
7265    Male              0     Yes        Yes      67
7266    Male              0     Yes        Yes      63

[7267 rows x 5 columns]

✅ Ahora eliminaremos los diccionarios del dataframe principal y agregaremos los datos normalizados

In [199]:
#Eliminamos las columnas con los diccionarios
df.drop(columns=['customer', 'phone', 'internet', 'account'], inplace=True)
#Concatenamos al dataframe las nuevas columnas
df = pd.concat([df, df_customer, df_phone, df_internet, df_account], axis=1)
df.sample(5)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
3842  5260-UMPWX    No  Female              0     Yes        Yes      26   
2335  3255-GRXMG    No    Male              0      No        Yes      18   
366   0526-SXDJP    No    Male              0     Yes         No      72   
1018  1429-UYJSV    No  Female              0      No         No      12   
1005  1407-DIGZV    No  Female              0     Yes        Yes       3   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
3842          Yes                No              No  ...  No internet service   
2335          Yes                No             DSL  ...                   No   
366            No  No phone service             DSL  ...                  Yes   
1018          Yes                No              No  ...  No internet service   
1005          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
3842  No internet service  No internet service  No internet service   
2335                   No                   No                   No   
366                   Yes                   No                   No   
1018  No internet service  No internet service  No internet service   
1005  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
3842  No internet service  Month-to-month              Yes   
2335                   No  Month-to-month              Yes   
366                    No        Two year               No   
1018  No internet service        One year               No   
1005  No internet service        Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
3842    Credit card (automatic)           20.25         493.95  
2335    Credit card (automatic)           45.65          747.2  
366   Bank transfer (automatic)           42.10           2962  
1018               Mailed check           19.70          260.9  
1005               Mailed check           19.10             52  

[5 rows x 21 columns]

✅ Ahora vemos que las columnas coinciden con el diccionario brindado

In [200]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


##2.2 Comprobación de incoherencias en los datos

In [201]:
#Verificamos duplicados
df.duplicated().sum()

np.int64(0)

In [202]:
#Verificamos valores nulos
df.isnull().sum()

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [203]:
#Verificamos valores únicos
pd.unique(df['Churn'])

array(['No', 'Yes', ''], dtype=object)

⚠️ OJO: Observamos que en `Churn` hay como valor único un String vacío, veamos cuántos hay en total

In [204]:
(df['Churn']=='').sum()

np.int64(224)

✅ Ahora veamos cuantos valores vacíos o espacios hay en las columnas

In [205]:
columnas = ['customerID', 'Churn', 'gender', 'SeniorCitizen', 'Partner',
       'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'Charges.Monthly',
       'Charges.Total']
cantidad_nulos = []
cantidad_espacios = []
for c in columnas:
  cantidad_nulos.append((df[c] == '').sum())
  cantidad_espacios.append((df[c] == ' ').sum())

df_nulos = pd.DataFrame({'columna': columnas, 'Nulos': cantidad_nulos, 'Espacios': cantidad_espacios})
df_nulos

columna  Nulos  Espacios
0         customerID      0         0
1              Churn    224         0
2             gender      0         0
3      SeniorCitizen      0         0
4            Partner      0         0
5         Dependents      0         0
6             tenure      0         0
7       PhoneService      0         0
8      MultipleLines      0         0
9    InternetService      0         0
10    OnlineSecurity      0         0
11      OnlineBackup      0         0
12  DeviceProtection      0         0
13       TechSupport      0         0
14       StreamingTV      0         0
15   StreamingMovies      0         0
16          Contract      0         0
17  PaperlessBilling      0         0
18     PaymentMethod      0         0
19   Charges.Monthly      0         0
20     Charges.Total      0        11

Se observa que hay **224** Strings nulos en la columna `Churn` y **11** Espacios en la columna `Charges.total`

Procedamos entonces a volverlos nulos para eliminarlos

In [206]:
#Reemplazamos esos datos por NaN
df['Churn'] = df['Churn'].replace('', np.nan)
df['Charges.Total'] = df['Charges.Total'].replace(' ', np.nan)

df.isnull().sum()

customerID            0
Churn               224
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod         0
Charges.Monthly       0
Charges.Total        11
dtype: int64

In [207]:
#Una vez vueltos nulos los eliminamos
df.dropna(inplace=True)
df[['Churn','Charges.Total']].isnull().sum()

Churn            0
Charges.Total    0
dtype: int64

##2.3 Manejo de inconsistencias

Verificamos si los datos tienen su correcto tipo

In [208]:
df.dtypes

customerID           object
Churn                object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
Charges.Monthly     float64
Charges.Total        object
dtype: object

❌ Sus Dtypes son incorrectos, vamos a corregirlos

In [209]:
#Datos tipo String
df['customerID'] = df['customerID'].str.strip() #para evitar que tengan espacios al inicio y al final
#Cambiemos a minusculas
df['gender'] = df['gender'].str.lower()
df['PaymentMethod'] = df['PaymentMethod'].str.lower()
df['Contract'] = df['Contract'].str.lower()

df[['customerID', 'gender', 'PaymentMethod', 'Contract']].sample(10)

customerID  gender              PaymentMethod        Contract
2273  3170-NMYVV  female  bank transfer (automatic)        two year
4278  5884-FBCTL  female  bank transfer (automatic)        two year
5060  6910-HADCM  female           electronic check  month-to-month
3524  4854-SSLTN    male               mailed check        two year
3901  5334-AFQJB    male    credit card (automatic)        two year
4878  6674-KVJHG  female           electronic check        one year
3190  4452-ROHMO  female               mailed check        two year
905   1270-XKUCC  female               mailed check        two year
5273  7198-GLXTC    male           electronic check  month-to-month
285   0411-EZJZE  female           electronic check  month-to-month

###2.3.1 Estandarización y transformación de datos

Para mayor entendimiento, cambiaremos los datos a español

In [210]:
df['gender'] = df['gender'].replace('male', 'masculino')
df['gender'] = df['gender'].replace('female', 'femenino')

df['gender'].sample(5)

3534    masculino
1120    masculino
5577    masculino
342      femenino
3552    masculino
Name: gender, dtype: object

In [211]:
df['PaymentMethod'].unique()

array(['mailed check', 'electronic check', 'credit card (automatic)',
       'bank transfer (automatic)'], dtype=object)

In [212]:
metodo_pago = {'mailed check':'cheque por correo',
               'electronic check':'cheque electrónico',
               'credit card (automatic)':'tarjeta de credito (automatica)',
               'bank transfer (automatic)':'transferencia bancaria (automatica)'}

df['PaymentMethod'] = df['PaymentMethod'].map(metodo_pago)
df['PaymentMethod'].sample(10)

7235                     cheque electrónico
4950        tarjeta de credito (automatica)
4162                     cheque electrónico
3173                      cheque por correo
4973        tarjeta de credito (automatica)
1895    transferencia bancaria (automatica)
4525        tarjeta de credito (automatica)
1030        tarjeta de credito (automatica)
470                      cheque electrónico
911                      cheque electrónico
Name: PaymentMethod, dtype: object

In [213]:
df['Contract'].unique()

array(['one year', 'month-to-month', 'two year'], dtype=object)

In [214]:
tipo_contrato = {
    'one year':'anual',
    'month-to-month':'mensual',
    'two year':'bianual'
}

df['Contract'] = df['Contract'].map(tipo_contrato)
df['Contract'].sample(10)

6894      anual
6628    bianual
3934      anual
3622    bianual
2807    mensual
2688      anual
29      bianual
2327    mensual
5399    mensual
2791      anual
Name: Contract, dtype: object

Ahora para los datos de `Yes` y `No` los volveremos booleanos

In [215]:
# Identificar columnas con Yes/No
yes_no_cols = [col for col in df.columns if df[col].isin(["Yes", "No"]).all()]

yes_no_cols

['Churn', 'Partner', 'Dependents', 'PhoneService', 'PaperlessBilling']

In [216]:
# Convertirlas a booleanos
df[yes_no_cols] = df[yes_no_cols].apply(lambda col: col.map({"Yes": True, "No": False}))

# Verificar cambios
df[yes_no_cols].head()

Churn  Partner  Dependents  PhoneService  PaperlessBilling
0  False     True        True          True              True
1  False    False       False          True             False
2   True    False       False          True              True
3   True     True       False          True              True
4   True     True       False          True              True

In [217]:
df[yes_no_cols].dtypes

Churn               bool
Partner             bool
Dependents          bool
PhoneService        bool
PaperlessBilling    bool
dtype: object

Terminemos de asignar los tipos de datos

In [218]:
#String
df['customerID'] = df['customerID'].astype('string')

#Integer
df['tenure'] = df['tenure'].astype('int64')

#Float
df['Charges.Monthly'] = df['Charges.Monthly'].astype('float64')
df['Charges.Total'] = df['Charges.Total'].astype('float64')

#cambiemos el tipo de dato de la columna "SeniorCitizen"
df['SeniorCitizen'] = df['SeniorCitizen'].map({0: False, 1: True})

#Category
categorias = ['gender', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaymentMethod']
df[categorias] = df[categorias].astype('category')

df.dtypes

customerID          string[python]
Churn                         bool
gender                    category
SeniorCitizen                 bool
Partner                       bool
Dependents                    bool
tenure                       int64
PhoneService                  bool
MultipleLines             category
InternetService           category
OnlineSecurity            category
OnlineBackup              category
DeviceProtection          category
TechSupport               category
StreamingTV               category
StreamingMovies           category
Contract                  category
PaperlessBilling              bool
PaymentMethod             category
Charges.Monthly            float64
Charges.Total              float64
dtype: object

✅ Por ultimo renombraremos las columnas a español

In [219]:
df = df.rename(columns={
    "customerID": "id_cliente",
    "Churn": "abandono",  # cliente se fue o no
    "gender": "genero",
    "SeniorCitizen": "adulto_mayor",
    "Partner": "pareja",
    "Dependents": "dependientes",
    "tenure": "meses_contrato",
    "PhoneService": "servicio_telefonico",
    "MultipleLines": "multiples_lineas",
    "InternetService": "servicio_internet",
    "OnlineSecurity": "seguridad_en_linea",
    "OnlineBackup": "respaldo_en_linea",
    "DeviceProtection": "proteccion_dispositivo",
    "TechSupport": "soporte_tecnico",
    "StreamingTV": "tv_streaming",
    "StreamingMovies": "peliculas_streaming",
    "Contract": "tipo_contrato",
    "PaperlessBilling": "factura_electronica",
    "PaymentMethod": "metodo_pago",
    "Charges.Monthly": "cargos_mensuales",
    "Charges.Total": "cargos_totales"
})

df.columns

Index(['id_cliente', 'abandono', 'genero', 'adulto_mayor', 'pareja',
       'dependientes', 'meses_contrato', 'servicio_telefonico',
       'multiples_lineas', 'servicio_internet', 'seguridad_en_linea',
       'respaldo_en_linea', 'proteccion_dispositivo', 'soporte_tecnico',
       'tv_streaming', 'peliculas_streaming', 'tipo_contrato',
       'factura_electronica', 'metodo_pago', 'cargos_mensuales',
       'cargos_totales'],
      dtype='object')

##2.4 Columna de cuentas diarias

In [221]:
df['cuentas_diarias'] = (df['cargos_mensuales']/30).round(2)

df['cuentas_diarias']

0       2.19
1       2.00
2       2.46
3       3.27
4       2.80
        ... 
7262    1.84
7263    2.84
7264    1.68
7265    2.26
7266    1.97
Name: cuentas_diarias, Length: 7032, dtype: float64

#📊 Carga y análisis

##3.1 Análisis Descriptivo

In [223]:
df.describe()

meses_contrato  cargos_mensuales  cargos_totales  cuentas_diarias
count     7032.000000       7032.000000     7032.000000      7032.000000
mean        32.421786         64.798208     2283.300441         2.159891
std         24.545260         30.085974     2266.771362         1.002955
min          1.000000         18.250000       18.800000         0.610000
25%          9.000000         35.587500      401.450000         1.187500
50%         29.000000         70.350000     1397.475000         2.340000
75%         55.000000         89.862500     3794.737500         2.992500
max         72.000000        118.750000     8684.800000         3.960000

#📄Informe final